File: frontend/lib/models/product.dart
import 'package:equatable/equatable.dart';

class Product extends Equatable {
  final int id;
  final String name;
  final String category;
  final int quantity;
  final double price;

  const Product({
    required this.id,
    required this.name,
    required this.category,
    required this.quantity,
    required this.price,
  });

  factory Product.fromJson(Map<String, dynamic> json) {
    return Product(
      id: json['id'] ?? 0,
      name: json['name'] ?? '',
      category: json['category'] ?? '',
      quantity: json['quantity'] ?? 0,
      price: (json['price'] ?? 0.0).toDouble(),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'name': name,
      'category': category,
      'quantity': quantity,
      'price': price,
    };
  }

  Product copyWith({
    int? id,
    String? name,
    String? category,
    int? quantity,
    double? price,
  }) {
    return Product(
      id: id ?? this.id,
      name: name ?? this.name,
      category: category ?? this.category,
      quantity: quantity ?? this.quantity,
      price: price ?? this.price,
    );
  }

  @override
  List<Object?> get props => [id, name, category, quantity, price];
} 

File: backend/models/product.go
package models

import (
	"time"

	"gorm.io/gorm"
)

type Product struct {
	ID        uint           `json:"id" gorm:"primaryKey"`
	Name      string         `json:"name" gorm:"not null;size:255" binding:"required"`
	Category  string         `json:"category" gorm:"not null;size:100" binding:"required"`
	Quantity  int            `json:"quantity" gorm:"not null;default:0" binding:"required,min=0"`
	Price     float64        `json:"price" gorm:"not null;type:decimal(10,2)" binding:"required,min=0"`
	CreatedAt time.Time      `json:"created_at"`
	UpdatedAt time.Time      `json:"updated_at"`
	DeletedAt gorm.DeletedAt `json:"-" gorm:"index"`
}

type CreateProductRequest struct {
	Name     string  `json:"name" binding:"required,min=2"`
	Category string  `json:"category" binding:"required"`
	Quantity int     `json:"quantity" binding:"required,min=0"`
	Price    float64 `json:"price" binding:"required,min=0"`
}

type UpdateProductRequest struct {
	Name     string  `json:"name" binding:"required,min=2"`
	Category string  `json:"category" binding:"required"`
	Quantity int     `json:"quantity" binding:"required,min=0"`
	Price    float64 `json:"price" binding:"required,min=0"`
}

type ProductFilter struct {
	Category    string `form:"category"`
	InStock     *bool  `form:"in_stock"`
	StockFilter string `form:"stock_filter"`
}
